### PR TITLE
Fix crash when using `RecordingStream::set_thread_local` on macOS 

### PR DESCRIPTION
**Pull request by emilk**
### What
* Closes https://github.com/rerun-io/rerun/issues/3558
* Closes https://github.com/rerun-io/rerun/issues/2889
* Opens https://github.com/rerun-io/rerun/issues/3937

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3929) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/3929)
- [Docs preview](https://rerun.io/preview/2b1469294de742102f69f16f6ecb8dc0e3e3a0c5/docs) 
- [Examples preview](https://rerun.io/preview/2b1469294de742102f69f16f6ecb8dc0e3e3a0c5/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: crates/re_sdk/src/global.rs [hunk 1]
```diff
@@ -28,14 +28,49 @@ impl std::fmt::Display for RecordingScope {
...
+impl Drop for ThreadLocalRecording {
+    fn drop(&mut self) {
+        if let Some(stream) = self.stream.take() {
+            // Work-around for https://github.com/rerun-io/rerun/issues/2889
+            // Calling drop on `self.stream` will panic the calling thread.
+            // But we want to make sure we don't loose the data in the stream.
+            // So how?
+            re_log::warn!("Using thread-local RecordingStream on macOS can result in data loss because of https://github.com/rerun-io/rerun/issues/2889");
+
+            // Give the batchet and sink threads a chance to process the data.
```
Comment (Wumpf) [inline crates/re_sdk/src/global.rs:57]:
```suggestion
            // Give the batcher and sink threads a chance to process the data.
```

File: crates/re_sdk/src/global.rs [hunk 2]
```diff
@@ -28,14 +28,49 @@ impl std::fmt::Display for RecordingScope {
...
+            std::thread::sleep(std::time::Duration::from_millis(500));
```
Comment (Wumpf) [inline crates/re_sdk/src/global.rs:58]: Tbh not sure if we already talked about this, but there's no way easy way right now to know whether the threads are still busy, right? Would be nice to be able to tell them to join instead 🤔 
Adding 500ms to our shutdown time is pretty awful and error prone.

**Review by Wumpf — APPROVED** [review]
_shudder_

File: crates/re_sdk/src/global.rs (hunk 2 above)
Comment (emilk) [reply crates/re_sdk/src/global.rs:58]: Yeah, it's not great, but we would need to check if all the queues are empty and the threads non-busy all the way down, and it is not trivial

Comment (emilk) [reply crates/re_sdk/src/global.rs:58]: I'll open a new issue (and close the old one) as a future TODO

Comment (emilk) [reply crates/re_sdk/src/global.rs:58]: https://github.com/rerun-io/rerun/issues/3937

**Pull request merged**